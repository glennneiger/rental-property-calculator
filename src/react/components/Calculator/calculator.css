@import "../../../styles.css";

.calculator {
  height: 100%;
  display: flex;
  flex-direction: column;
  background-color: var(--mainBackgroundColor);
  padding: 12px 0px;
}

.divider {
  width: 97%;
  margin: 0 auto;
  height: 2px;
  background-color: rgba(0, 0, 0, 0.8);
}

@media(--desktop-up) {
  .calculator {
    padding-left: var(--calculatorContentPadding);
    padding-right: var(--calculatorContentPadding);
  }

  .calculatorWithSidebar {
    padding-left: 0px;
  }
}