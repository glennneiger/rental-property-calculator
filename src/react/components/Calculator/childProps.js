import {
  CALCULATOR_INPUT_WIDTH_DEFAULT,
  INPUT_DESCRIPTION_DOWN_PAYMENT,
  INPUT_DESCRIPTION_LOAN_AMOUNT,
  INPUT_DESCRIPTION_PURCHASE_PRICE,
  INPUT_DESCRIPTION_AFTER_REPAIR_VALUE,
  INPUT_DESCRIPTION_CLOSING_COSTS,
  INPUT_DESCRIPTION_INTEREST_RATE,
  INPUT_DESCRIPTION_AMORTIZATION_PERIOD,
  INPUT_DESCRIPTION_OTHER_INCOME,
  INPUT_DESCRIPTION_OTHER_INITIAL_COSTS,
  INPUT_DESCRIPTION_RENTAL_INCOME,
  INPUT_DESCRIPTION_REPAIR_COSTS,
  INPUT_ID_AFTER_REPAIR_VALUE,
  INPUT_ID_AMORTIZATION_PERIOD,
  INPUT_ID_ANNUAL_CONSTANT_EXPENSES_GROWTH,
  INPUT_ID_ANNUAL_INCOME_GROWTH,
  INPUT_ID_CAP_EX,
  INPUT_ID_CITY,
  INPUT_ID_CLOSING_COSTS,
  INPUT_ID_COUNTRY,
  INPUT_ID_DOWN_PAYMENT,
  INPUT_ID_ELECTRICITY,
  INPUT_ID_GARBAGE,
  INPUT_ID_HOA,
  INPUT_ID_INSURANCE,
  INPUT_ID_INTEREST_RATE,
  INPUT_ID_LOAN_AMOUNT,
  INPUT_ID_MANAGEMENT,
  INPUT_ID_MORTGAGE,
  INPUT_ID_NEIGHBORHOOD,
  INPUT_ID_OTHER_EXPENSES,
  INPUT_ID_OTHER_INCOME,
  INPUT_ID_OTHER_INITIAL_COSTS,
  INPUT_ID_POSTAL_OR_ZIP_CODE,
  INPUT_ID_PRIVATE_MORTGAGE_INSURANCE,
  INPUT_ID_PROPERTY_TAX,
  INPUT_ID_PROPERTY_VALUE_GROWTH,
  INPUT_ID_PROVINCE_OR_STATE,
  INPUT_ID_PURCHASE_PRICE,
  INPUT_ID_RENTAL_INCOME,
  INPUT_ID_REPAIR_COSTS,
  INPUT_ID_REPAIRS_AND_MAINTENANCE,
  INPUT_ID_SQUARE_FEET,
  INPUT_ID_STREET_ADDRESS,
  INPUT_ID_UNIT_NUMBER,
  INPUT_ID_VACANCY,
  INPUT_ID_WATER_AND_SEWER,
  LABEL_AFTER_REPAIR_VALUE,
  LABEL_AMORTIZATION_PERIOD,
  LABEL_CAP_EX,
  LABEL_CITY,
  LABEL_CLOSING_COSTS,
  LABEL_CONSTANT_EXPENSES_GROWTH,
  LABEL_COUNTRY,
  LABEL_DOWN_PAYMENT,
  LABEL_ELECTRICITY,
  LABEL_GARBAGE,
  LABEL_HOA,
  LABEL_INCOME_GROWTH,
  LABEL_INSURANCE,
  LABEL_INTEREST_RATE,
  LABEL_LOAN_AMOUNT,
  LABEL_MANAGEMENT,
  LABEL_MORTGAGE,
  LABEL_NEIGHBORHOOD,
  LABEL_OTHER_EXPENSES,
  LABEL_OTHER_INCOME,
  LABEL_OTHER_INITIAL_COSTS,
  LABEL_POSTAL_OR_ZIP_CODE,
  LABEL_PRIVATE_MORTGAGE_INSURANCE,
  LABEL_PROPERTY_TAX,
  LABEL_PROPERTY_VALUE_GROWTH,
  LABEL_PROVINCE_OR_STATE,
  LABEL_PURCHASE_PRICE,
  LABEL_RENTAL_INCOME,
  LABEL_REPAIR_COSTS,
  LABEL_REPAIRS_AND_MAINTENANCE,
  LABEL_SQUARE_FEET,
  LABEL_STREET_ADDRESS,
  LABEL_UNIT_NUMBER,
  LABEL_VACANCY,
  LABEL_WATER_AND_SEWER,
  TITLE_FUTURE_PROJECTIONS,
  TITLE_GENERAL_INFO,
  TITLE_INITIAL_PURCHASE,
  TITLE_MONTHLY_EXPENSES,
  TITLE_MONTHLY_INCOME
} from '../../../constants';

const generalInfoInputProps = [
  {
    inputId: INPUT_ID_NEIGHBORHOOD,
    inputType: 'text',
    label: LABEL_NEIGHBORHOOD,
    textInputWidth: 200
  },
  {
    inputId: INPUT_ID_STREET_ADDRESS,
    inputType: 'text',
    label: LABEL_STREET_ADDRESS,
    textInputWidth: 210
  },
  {
    inputId: INPUT_ID_UNIT_NUMBER,
    inputType: 'text',
    label: LABEL_UNIT_NUMBER,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_CITY,
    inputType: 'text',
    label: LABEL_CITY,
    textInputWidth: 200
  },
  {
    inputId: INPUT_ID_PROVINCE_OR_STATE,
    inputType: 'text',
    label: LABEL_PROVINCE_OR_STATE,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_COUNTRY,
    inputType: 'text',
    label: LABEL_COUNTRY,
    textInputWidth: 200
  },
  {
    inputId: INPUT_ID_POSTAL_OR_ZIP_CODE,
    inputType: 'text',
    label: LABEL_POSTAL_OR_ZIP_CODE,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_SQUARE_FEET,
    inputType: 'number',
    label: LABEL_SQUARE_FEET,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  }
];

const initialPurchaseInputProps = [
  {
    inputDescription: INPUT_DESCRIPTION_PURCHASE_PRICE,
    inputId: INPUT_ID_PURCHASE_PRICE,
    inputType: 'number',
    label: LABEL_PURCHASE_PRICE,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputDescription: INPUT_DESCRIPTION_DOWN_PAYMENT,
    inputId: INPUT_ID_DOWN_PAYMENT,
    inputType: 'number',
    label: LABEL_DOWN_PAYMENT,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputDescription: INPUT_DESCRIPTION_LOAN_AMOUNT,
    inputId: INPUT_ID_LOAN_AMOUNT,
    inputType: 'number',
    label: LABEL_LOAN_AMOUNT,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputDescription: INPUT_DESCRIPTION_INTEREST_RATE,
    inputId: INPUT_ID_INTEREST_RATE,
    inputType: 'number',
    label: LABEL_INTEREST_RATE,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputDescription: INPUT_DESCRIPTION_AMORTIZATION_PERIOD,
    inputId: INPUT_ID_AMORTIZATION_PERIOD,
    inputType: 'number',
    label: LABEL_AMORTIZATION_PERIOD,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputDescription: INPUT_DESCRIPTION_AFTER_REPAIR_VALUE,
    inputId: INPUT_ID_AFTER_REPAIR_VALUE,
    inputType: 'number',
    label: LABEL_AFTER_REPAIR_VALUE,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputDescription: INPUT_DESCRIPTION_REPAIR_COSTS,
    inputId: INPUT_ID_REPAIR_COSTS,
    inputType: 'number',
    label: LABEL_REPAIR_COSTS,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputDescription: INPUT_DESCRIPTION_CLOSING_COSTS,
    inputId: INPUT_ID_CLOSING_COSTS,
    inputType: 'number',
    label: LABEL_CLOSING_COSTS,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputDescription: INPUT_DESCRIPTION_OTHER_INITIAL_COSTS,
    inputId: INPUT_ID_OTHER_INITIAL_COSTS,
    inputType: 'number',
    label: LABEL_OTHER_INITIAL_COSTS,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  }
];

/* Only add stuff here to calculate income */
export const incomeInputProps = [
  {
    inputDescription: INPUT_DESCRIPTION_RENTAL_INCOME,
    inputId: INPUT_ID_RENTAL_INCOME,
    inputType: 'number',
    label: LABEL_RENTAL_INCOME,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputDescription: INPUT_DESCRIPTION_OTHER_INCOME,
    inputId: INPUT_ID_OTHER_INCOME,
    inputType: 'number',
    label: LABEL_OTHER_INCOME,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  }
];

/* Only add stuff here to calculate expenses */
export const expensesInputProps = [
  {
    inputId: INPUT_ID_MORTGAGE,
    inputType: 'number',
    label: LABEL_MORTGAGE,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_ELECTRICITY,
    inputType: 'number',
    label: LABEL_ELECTRICITY,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_WATER_AND_SEWER,
    inputType: 'number',
    label: LABEL_WATER_AND_SEWER,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_PRIVATE_MORTGAGE_INSURANCE,
    inputType: 'number',
    label: LABEL_PRIVATE_MORTGAGE_INSURANCE,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_GARBAGE,
    inputType: 'number',
    label: LABEL_GARBAGE,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_HOA,
    inputType: 'number',
    label: LABEL_HOA,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_INSURANCE,
    inputType: 'number',
    label: LABEL_INSURANCE,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_PROPERTY_TAX,
    inputType: 'number',
    label: LABEL_PROPERTY_TAX,
    percentOfPropertyValue: true,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_VACANCY,
    inputType: 'number',
    label: LABEL_VACANCY,
    percentOfRent: true,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_REPAIRS_AND_MAINTENANCE,
    inputType: 'number',
    label: LABEL_REPAIRS_AND_MAINTENANCE,
    percentOfRent: true,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_CAP_EX,
    inputType: 'number',
    label: LABEL_CAP_EX,
    percentOfRent: true,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_MANAGEMENT,
    inputType: 'number',
    label: LABEL_MANAGEMENT,
    percentOfRent: true,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_OTHER_EXPENSES,
    inputType: 'number',
    label: LABEL_OTHER_EXPENSES,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  }
];

export const futureProjectionsInputProps = [
  {
    inputId: INPUT_ID_PROPERTY_VALUE_GROWTH,
    inputType: 'number',
    label: LABEL_PROPERTY_VALUE_GROWTH,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_ANNUAL_INCOME_GROWTH,
    inputType: 'number',
    label: LABEL_INCOME_GROWTH,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  },
  {
    inputId: INPUT_ID_ANNUAL_CONSTANT_EXPENSES_GROWTH,
    inputType: 'number',
    label: LABEL_CONSTANT_EXPENSES_GROWTH,
    textInputWidth: CALCULATOR_INPUT_WIDTH_DEFAULT
  }
];

export const inputSectionData = [
  {
    title: TITLE_GENERAL_INFO,
    childProps: generalInfoInputProps
  },
  {
    title: TITLE_INITIAL_PURCHASE,
    childProps: initialPurchaseInputProps
  },
  {
    title: TITLE_MONTHLY_INCOME,
    childProps: incomeInputProps
  },
  {
    title: TITLE_MONTHLY_EXPENSES,
    childProps: expensesInputProps
  },
  {
    title: TITLE_FUTURE_PROJECTIONS,
    childProps: futureProjectionsInputProps
  }
];
