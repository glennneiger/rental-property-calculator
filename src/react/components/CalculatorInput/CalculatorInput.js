import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import css from './calculatorInput.css';
import TextInput from '../TextInput';
import { INPUT_ID_AMORTIZATION_PERIOD } from '../../../constants';

class CalculatorInput extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    const {
      inputId,
      inputType,
      section,
      setChangesMade,
      updateInput
    } = this.props;

    const value = event.target.value;
    if (inputType === 'number' && !value.match(/^\d*\.?\d{0,2}$/)) {
      return;
    }
    if (inputId === INPUT_ID_AMORTIZATION_PERIOD && value.length > 3) {
      return;
    }
    updateInput(value, section, inputId);
    setChangesMade(true);
  }

  render() {
    const {
      content,
      inputId,
      label,
      sidebarVisible,
      textInputWidth
    } = this.props;
    return (
      <div className={classNames({
        [css.calculatorInput]: true,
        [css.calculatorInputNoSidebar]: !sidebarVisible,
        [css.calculatorInputWithSidebar]: sidebarVisible
      })}>
        <label htmlFor={inputId}>{label}</label>
        <TextInput type={'text'}
          id={inputId}
          style={{ width: textInputWidth }}
          value={content}
          onChange={this.handleChange}
          width={textInputWidth}
        />
      </div>
    );
  }
}

CalculatorInput.propTypes = {
  content: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  inputId: PropTypes.string.isRequired,
  inputType: PropTypes.string,
  label: PropTypes.string.isRequired,
  section: PropTypes.string.isRequired,
  setChangesMade: PropTypes.func.isRequired,
  sidebarVisible: PropTypes.bool.isRequired,
  textInputWidth: PropTypes.number,
  updateInput: PropTypes.func.isRequired
};

export default CalculatorInput;