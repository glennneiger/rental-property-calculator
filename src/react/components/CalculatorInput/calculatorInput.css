@import "../../../styles.css";

.calculatorInput {
  display: flex;
  flex-direction: column;
  padding: 18px 12px;
  font-size: 1.2em;
  width: 100%;
  font-family: 'Raleway', sans-serif;
}

.calculatorInput {
  & label {
    font-size: 0.8em;
    font-weight: bold;
    margin-bottom: 4px;
  }
}

@media (min-width: 650px) {
  .calculatorInputNoSidebar {
    width: 50%;
  }
}

@media (min-width: 850px) {
  .calculatorInputWithSidebar {
    width: 50%;
  }
}

@media (min-width: 1024px) {
  .calculatorInput {
    width: 33%;
  }
}

@media (--desktop-up) {
  .calculatorInput {
    width: 33%;
  }
}

@media (--large-up) {
  .calculatorInput {
    width: 20%;
  }
}