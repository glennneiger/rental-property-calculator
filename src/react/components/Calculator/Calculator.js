import React from 'react';
import PropTypes from 'prop-types';

import Result from '../Result';
import CalculatorInputSection from '../CalculatorInputSection';
import CalculatorInput from '../CalculatorInput';
import {
  inputSectionData
} from './childProps';
import css from './calculator.css';
import classNames from 'classnames';

const Calculator = ({
  sidebarVisible
}) => (
  <div className={classNames({
    [css.calculator]: true,
    [css.calculatorWithSidebar]: sidebarVisible
  })}>
    {inputSectionData.map(section => (
      <div key={section.title}>
        <CalculatorInputSection
          title={section.title}>
          {section.childProps.map(props => (
            <CalculatorInput
              key={props.inputId}
              {...props}
              section={section.title}
            />
          ))}
        </CalculatorInputSection>
        {<div className={css.divider} />}
      </div>
    ))}
    <Result />
  </div>
);

Calculator.propTypes = {
  sidebarVisible: PropTypes.bool.isRequired
};

export default Calculator;