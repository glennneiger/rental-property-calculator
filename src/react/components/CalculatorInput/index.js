import { connect } from 'react-redux';

import CalculatorInput from './CalculatorInput';
import {
  updateInput
} from '../../../actions/calculatorFields';
import {
  setChangesMade
} from '../../../actions/currentCalculation';

export const mapStateToProps = (state, ownProps) => ({
  content: state.calculator[ownProps.section][ownProps.inputId],
  sidebarVisible: state.sidebarVisible
});

const mapDispatchToProps = dispatch => ({
  updateInput(value, section, inputId) {
    dispatch(updateInput(value, section, inputId));
  },
  setChangesMade(changesMade) {
    dispatch(setChangesMade(changesMade));
  }
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CalculatorInput);