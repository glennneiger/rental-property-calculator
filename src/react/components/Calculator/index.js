import { connect } from 'react-redux';

import Calculator from './Calculator';

export const mapStateToProps = state => ({
  sidebarVisible: state.sidebarVisible
});

export default connect(mapStateToProps)(Calculator);