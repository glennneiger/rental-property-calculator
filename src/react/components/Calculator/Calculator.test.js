import React from 'react';
import { shallow } from 'enzyme';
import { expect as chaiExpect } from 'chai';
import ShallowRenderer from 'react-test-renderer/shallow';
import * as childProps from './childProps';
import CalculatorInput from '../CalculatorInput';

import Calculator from './Calculator';

const originalInputSectionData = childProps.inputSectionData;

describe('<Calculator />', () => {
  it('has proper className for styling', () => {
    const wrapper = shallow(<Calculator />);

    chaiExpect(wrapper).to.have.className('calculator');
  });

  it('renders properly with current inputSectionData', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<Calculator />);

    const tree = renderer.getRenderOutput();

    expect(tree).toMatchSnapshot();
  });

  it('renders the correct number of CalculatorInputSections', () => {
    const wrapper = shallow(<Calculator />);

    chaiExpect(wrapper.find('CalculatorInputSection'))
      .to.have.length(childProps.inputSectionData.length);
  });

  it('renders correct number of CalculatorInputs', () => {
    const mockTitle = 'General Info';
    const mockInputProps = [
      {
        inputId: 1,
        inputType: 'text',
        label: 'Neighborhood',
        textInputWidth: 200
      },
      {
        inputId: 2,
        inputType: 'text',
        label: 'Street Address',
        textInputWidth: 210
      }
    ];

    childProps.inputSectionData = [
      {
        title: mockTitle,
        childProps: mockInputProps
      }
    ];

    const wrapper = shallow(<Calculator />);

    chaiExpect(wrapper.find(CalculatorInput))
      .to.have.length(mockInputProps.length);

    childProps.inputSectionData = originalInputSectionData;
  });
});