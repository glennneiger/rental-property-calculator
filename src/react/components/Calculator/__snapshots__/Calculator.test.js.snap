// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator /> renders properly with current inputSectionData 1`] = `
<div
  className="calculator calculatorWithSidebar"
>
  <div>
    <CalculatorInputSection
      title="General Info"
    >
      <Connect(CalculatorInput)
        inputId="neighborhoodInput"
        inputType="text"
        label="Neighborhood"
        section="General Info"
        textInputWidth={200}
      />
      <Connect(CalculatorInput)
        inputId="streetAddressInput"
        inputType="text"
        label="Street Address"
        section="General Info"
        textInputWidth={210}
      />
      <Connect(CalculatorInput)
        inputId="unitNumberInput"
        inputType="text"
        label="Unit Number"
        section="General Info"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="cityInput"
        inputType="text"
        label="City"
        section="General Info"
        textInputWidth={200}
      />
      <Connect(CalculatorInput)
        inputId="provinceOrStateInput"
        inputType="text"
        label="Province/State"
        section="General Info"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="countryInput"
        inputType="text"
        label="Country"
        section="General Info"
        textInputWidth={200}
      />
      <Connect(CalculatorInput)
        inputId="postalOrZipCodeInput"
        inputType="text"
        label="Postal/Zip Code"
        section="General Info"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="squareFeetInput"
        inputType="number"
        label="Square Feet"
        section="General Info"
        textInputWidth={160}
      />
    </CalculatorInputSection>
    <div
      className="divider"
    />
  </div>
  <div>
    <CalculatorInputSection
      title="Initial Purchase"
    >
      <Connect(CalculatorInput)
        inputId="purchasePriceInput"
        inputType="number"
        label="Purchase Price"
        section="Initial Purchase"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="downPaymentInput"
        inputType="number"
        label="Down Payment"
        section="Initial Purchase"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="loanAmountInput"
        inputType="number"
        label="Loan Amount"
        section="Initial Purchase"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="interestRateInput"
        inputType="number"
        label="Interest Rate (%)"
        section="Initial Purchase"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="amortizationPeriodInput"
        inputType="number"
        label="Amortization Period (years)"
        section="Initial Purchase"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="afterRepairValueInput"
        inputType="number"
        label="After Repair Value"
        section="Initial Purchase"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="repairCostsInput"
        inputType="number"
        label="Repair Costs"
        section="Initial Purchase"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="closingCostsInput"
        inputType="number"
        label="Closing Costs"
        section="Initial Purchase"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="otherInitialCostsInput"
        inputType="number"
        label="Other Initial Costs"
        section="Initial Purchase"
        textInputWidth={160}
      />
    </CalculatorInputSection>
    <div
      className="divider"
    />
  </div>
  <div>
    <CalculatorInputSection
      title="Monthly Income"
    >
      <Connect(CalculatorInput)
        inputId="rentalIncomeInput"
        inputType="number"
        label="Rental Income"
        section="Monthly Income"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="otherIncomeInput"
        inputType="number"
        label="Other Income"
        section="Monthly Income"
        textInputWidth={160}
      />
    </CalculatorInputSection>
    <div
      className="divider"
    />
  </div>
  <div>
    <CalculatorInputSection
      title="Monthly Expenses"
    >
      <Connect(CalculatorInput)
        inputId="mortgageInput"
        inputType="number"
        label="Mortgage"
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="electricityInput"
        inputType="number"
        label="Electricity"
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="waterAndSewerInput"
        inputType="number"
        label="Water and Sewer"
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="privateMortgageInsuranceInput"
        inputType="number"
        label="Private Mortgage Insurance"
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="garbageInput"
        inputType="number"
        label="Garbage"
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="hoaInput"
        inputType="number"
        label="HOA Fees"
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="insuranceInput"
        inputType="number"
        label="Insurance"
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="propertyTaxInput"
        inputType="number"
        label="Property Tax (%)"
        percentOfPropertyValue={true}
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="vacancyInput"
        inputType="number"
        label="Vacancy (%)"
        percentOfRent={true}
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="r&mInput"
        inputType="number"
        label="Repairs & Maintenance (%)"
        percentOfRent={true}
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="capExInput"
        inputType="number"
        label="Cap. Ex. (%)"
        percentOfRent={true}
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="managementInput"
        inputType="number"
        label="Management (%)"
        percentOfRent={true}
        section="Monthly Expenses"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="otherExpensesInput"
        inputType="number"
        label="Other Expenses"
        section="Monthly Expenses"
        textInputWidth={160}
      />
    </CalculatorInputSection>
    <div
      className="divider"
    />
  </div>
  <div>
    <CalculatorInputSection
      title="Future Projections"
    >
      <Connect(CalculatorInput)
        inputId="propertyValueGrowthInput"
        inputType="number"
        label="Annual Property Value Growth (%)"
        section="Future Projections"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="annualIncomeGrowthInput"
        inputType="number"
        label="Annual Income Growth (%)"
        section="Future Projections"
        textInputWidth={160}
      />
      <Connect(CalculatorInput)
        inputId="annualExpensesGrowthInput"
        inputType="number"
        label="Annual Constant Expenses Growth (%)"
        section="Future Projections"
        textInputWidth={160}
      />
    </CalculatorInputSection>
    <div
      className="divider"
    />
  </div>
  <Connect(Result) />
</div>
`;
